spyder/plugins/plots/widgets/main_widget.py

```diff
@@ -0,0 +1,468 @@
+        """
+        Register shell with figure explorer.
+
+        This function opens a new FigureBrowser for browsing the figures
+        in the shell.
+        """
+        shellwidget_id = id(shellwidget)
+        if shellwidget_id not in self._shellwidgets:
+            fig_browser = FigureBrowser(parent=self._stack,
+                                        background_color=MAIN_BG_COLOR)
+            fig_browser.set_shellwidget(shellwidget)
+            fig_browser.sig_redirect_stdio_requested.connect(
+                self.sig_redirect_stdio_requested)
+            self.add_widget(fig_browser)
+            self._shellwidgets[shellwidget_id] = fig_browser
+            self.set_shellwidget(shellwidget)
+            return fig_browser
+
+    def remove_shellwidget(self, shellwidget):
+        # If shellwidget_id is not in self._shellwidgets, it simply means
+        # that shell was not a Python-based console (it was a terminal)
+        shellwidget_id = id(shellwidget)
+        if shellwidget_id in self._shellwidgets:
+            fig_browser = self._shellwidgets.pop(shellwidget_id)
+            self.remove_widget(fig_browser)
+            fig_browser.close()
+
+    def set_shellwidget(self, shellwidget):
+        shellwidget_id = id(shellwidget)
+        if shellwidget_id in self._shellwidgets:
+            fig_browser = self._shellwidgets[shellwidget_id]
+            fig_browser.setup(self._options)
+            self.set_current_widget(fig_browser)
+
+    # --- API: Main widget
+    # ------------------------------------------------------------------------
+    def show_figure_menu(self, qpoint):
+        self._right_clicked_thumbnail = None
+        widget = self.current_widget()
+        if widget:
+            self.get_menu(PluginMainWidgetMenus.Context).popup(qpoint)
+
+    def show_thumbnail_menu(self, qpoint, thumbnail):
+        self._right_clicked_thumbnail = thumbnail
+        widget = self.current_widget()
+        if widget:
+            self.get_menu(PluginMainWidgetMenus.Context).popup(qpoint)
+
+    def save_plot(self):
+        widget = self._stack.currentWidget()
+        if widget:
+            if self._right_clicked_thumbnail is None:
+                widget.thumbnails_sb.save_current_figure_as()
+            else:
+                widget.thumbnails_sb.save_thumbnail_figure_as(
+                    self._right_clicked_thumbnail)
+
+    def save_all_plots(self):
+        widget = self._stack.currentWidget()
+        if widget:
+            widget.thumbnails_sb.save_all_figures_as()
+
+    def copy_image(self):
+        widget = self._stack.currentWidget()
+        if widget and widget.figviewer and widget.figviewer.figcanvas.fig:
+            if self._right_clicked_thumbnail is None:
+                widget.figviewer.figcanvas.copy_figure()
+            else:
+                self._right_clicked_thumbnail.canvas.copy_figure()
+
+    def remove_plot(self):
+        widget = self._stack.currentWidget()
```

Comment:
Missing docstring and `self.current_widget()`

Reply:
Fixed